src/main/java/algorithm/VisualizationManager.java

```diff
@@ -0,0 +1,395 @@
+     *
+     * @param meshGroup new meshGroup
+     */
+    public void setMeshGroup(Group meshGroup) {
+        this.meshGroup = meshGroup;
+    }
+
+//    public void setViewportSize(double VIEWPORT_SIZE) {
+//        this.VIEWPORT_SIZE = VIEWPORT_SIZE;
+//        this.VIEWPORT_CENTER = VIEWPORT_SIZE / 2;
+//    }
+
+
+    /**
+     * Creates a Tracking Cone for each device
+     */
+    private void createTrackingCone() {
+        if (trackingService.getTrackingDataSource() == null) {
+            statusLabel.setText("No Tracking Data Source");
+            return;
+        }
+        statusLabel.setText("");
+
+        // update after CSV/IGT data has been loaded
+        trackingService.getTrackingDataSource().update();
+        ArrayList<Tool> tools = trackingService.getTrackingDataSource().getLastToolList();
+
+        if (tools != null) {
+            trackingCones = new ConeMesh[tools.size()];
+
+            for (int i = 0; i < trackingCones.length; i++) {
+                trackingCones[i] = new ConeMesh(12, 4, 8);
+            }
+        }
+    }
+
+    /**
+     * loads an STL File for Visualization
+     */
+    public void loadStlModel() {
+        StlMeshImporter importer = new StlMeshImporter();
+        FileChooser fc = new FileChooser();
+        fc.setTitle("Load STL File");
+        fc.getExtensionFilters().add(new FileChooser.ExtensionFilter("STL Files", "*.stl"));
+        List<File> fileList = fc.showOpenMultipleDialog(new Stage());
+
+        if (fileList != null) {
+            stlFiles = new MeshView[fileList.size()];
+            int x = 0;
+
+            for (File file : fileList) {
+                try {
+                    importer.read(file);
+                    Mesh mesh = importer.getImport();
+                    stlFiles[x] = new MeshView(mesh);
+
+                    x++;
+
+                } catch (Exception e) {
+                    logger.log(Level.WARNING, "Error reading STL file");
+                }
+            }
+        }
+    }
+
+    /**
+     * Start the visualization of tracking data that is read from CSV or IGT connection
+     */
+    public void startVisualization() {
+        if (trackingService.getTrackingDataSource() == null) {
+            statusLabel.setText("No Tracking Data Source");
+            return;
+        }
+        if (stlFiles == null) {
+            statusLabel.setText("Load STL Model for Visualization");
+            return;
+        }
+        // timeline has not been started in trackingData view
+        if (trackingService.getTimeline() == null) {
+            statusLabel.setText("Start Tracking in Tracking Data View first");
+            return;
+        }
+        statusLabel.setText("");
+
+        // loads the next set of tracking data
+        trackingService.getTrackingDataSource().update();
+        List<ToolMeasure> tools = trackingService.getDataService().loadNextData(1);
+
+        int j = 0;
+
+        for (ToolMeasure tool : tools) {
```

Comment:
Why use a foreach-loop if you could do a fori-loop to directly get an index variable? :)

Reply:
absolutely right, changed it